c10/core/TensorImpl.cpp

```diff
@@ -57,7 +58,7 @@ TensorImpl::TensorImpl(Storage&& storage, TensorTypeSet type_set, const caffe2::
       numel_(0),
       data_type_(data_type),
       device_opt_(device_opt),
-      type_set_(type_set) {
+      type_set_(type_set.remove(TensorTypeId::VariableTensorId)) {
```

Comment:
Yeah, it's pretty goofy. Basically, the invariant is `type_set.has(TensorTypeId::VariableTensorId) iff autograd_meta_ != nullptr`, right? Well, on construction, `autograd_meta_` is always null. So to preserve the invariant we have to remove variable tensor id if it was passed in (which it might be, if you are doing a shallow copy from a variable thing).

@gchanan mentioned earlier that we could make things a bit better if we moved some of the type set logic into the variable constructors. If we can move `autograd_meta_` initialization into these constructors, we can make this a bit more transparent. But that involves a lot more moving code around. In any case, we probably need a Note for this, I'll go add one.

Reply:
Yeah, I think it is fine, especially for now. Ideally this would assert if variable was set, and we'd take care not to shallow initialize something with variable without also setting autograd_meta_ at the same time. I think this is similar or equivalent to what you were saying in your comment.